Scripted tests

britter opened the pull request:
My initial approach towards adding scripted tests. Still WIP, needs some polishing. I'm looking for feedback whether this is going in the right direction. I've created the tests according to http://eed3si9n.com/testing-sbt-plugins

britter commented:
Looks like at least Travis is happy with these changes :)

britter commented on scripted.sbt:
```diff
@@ -0,0 +1,7 @@
+ScriptedPlugin.scriptedSettings
+
+scriptedLaunchOpts := { scriptedLaunchOpts.value ++
+  Seq("-Xmx1024M", "-XX:MaxPermSize=256M", "-Dplugin.version=" + version.value)
```
`MaxPermSize` has been removed in Java 8. Can we drop this?

hseeberger replied:
```
scriptedSettings
scriptedLaunchOpts := scriptedLaunchOpts.value ++ Vector(
  "-Xmx1024M",
  s"-Dplugin.version=${version.value}"
)
```

britter commented on src/sbt-test/sbt-header/simple/build.sbt:
```diff
@@ -0,0 +1,21 @@
+  "scala" -> Apache2_0("2015", "Heiko Seeberger")
+)
+
+TaskKey[Unit]("check") <<= scalaSource in Compile map { (scalaSource) =>
```
Is there a better way to compare file contents? Maybe some sbt API I'm not aware of?

hseeberger replied:
Nothing I'm aware of.

hseeberger replied:
Maybe rename `check` to `checkFileContents`?

britter commented on src/sbt-test/sbt-header/simple/src/main/scala/Main.scala_expected:
```diff
@@ -0,0 +1,19 @@
+
+package de.heikoseeberger.sbtheader.test;
+
+class Main
```
My idea was to have an actual file and an expected file. Is that a good idea? Or do we want to build the expectations from the actual file in our `check` task?

hseeberger replied:
I like your approach, but as mentioned above, the file should probably go under `resources`.

britter replied:
Putting the expected files alongside the actual files makes it easier to select them inside the check task. But I agree, src/main/scala should only contain scala files. So which way do we go? Ease of use or clean project structure?

hseeberger replied:
Always go for simplicity, never for easiness. In this case, a resource should simply go under the resource directory.

britter replied:
In this case I see the simplicity in the `checkFileContents` task, because I only need to know the value of `scalaSource`. When I move the expected files to the resources directory, `checkFileContents` has to know both locations.

hseeberger replied:
Don't confuse simplicity with easiness!

hseeberger commented on .travis.yml:
```diff
@@ -1,4 +1,6 @@
 - oraclejdk8
-script: sbt +test
+script: 
+  - sbt +test
```
For sbt we always only use one Scala version, hence you can remove the `+`

britter replied:
Done! Note that the standard Travis build command is `sbt ++$TRAVIS_SCALA_VERSION test`. Maybe we should that instead?

hseeberger replied:
Not sure why we need special support from Travis here?

britter replied:
Because otherwise we cannot use the following in the future:

```yaml
   - 2.9.3
   - 2.10.4
   - 2.11.2
```

This is the reason I tend to restore the original Travis command when I add additional build steps.

hseeberger replied:
I don't want to redundantly specify the Scala version for Travis. All we need is `crossScalaVersions` in the build and the use the `+`.

hseeberger commented on project/scripted.sbt:
```diff
@@ -0,0 +1,3 @@
+libraryDependencies <+= sbtVersion { sv =>
```
Use `libraryDependencies += "org.scala-sbt" % "scripted-plugin" % sbtVersion.value` (see sbt-fresh).

hseeberger commented on scripted.sbt:
```diff
@@ -0,0 +1,7 @@
+  Seq("-Xmx1024M", "-XX:MaxPermSize=256M", "-Dplugin.version=" + version.value)
+}
+
+scriptedBufferLog := false
```
Why that?

britter replied:
Copy pasted from http://eed3si9n.com/testing-sbt-plugins :-)

I checked the build without this option. The build log will not print as much information then. I'd recommend we keep it.

hseeberger replied:
Fine!

hseeberger commented on src/sbt-test/sbt-header/simple/build.sbt:
```diff
@@ -0,0 +1,21 @@
+import de.heikoseeberger.sbtheader.license.Apache2_0
```
Please use the names I am using in sbt-fresh:
  - call the build file `test.sbt`
  - call the task according to what it is doing
  - also use the task marco
  - don't use outdated operators (I guess you have learned from Eugene's old blog post)

Here is the example:
```
import org.eclipse.jgit.api.Git
import org.eclipse.jgit.storage.file.FileRepositoryBuilder
import scala.collection.JavaConverters._

val checkInitialCommit = taskKey[Unit]("Verify the initial commit")
checkInitialCommit := {
  val git = new Git((new FileRepositoryBuilder).setWorkTree(baseDirectory.value).build())
  val commits = git.log().call().asScala.toVector
  assert(commits.size == 1, "There must be exactly one commit!")
}
```

britter replied:
I've tried to get this working, but I don't understand how to access `scalaSource` when I simply use `scalaSource.value` inside the task I get 

```
[info] [error] References to undefined settings:
[info] [error]
[info] [error]   *:scalaSource from *:checkFileContents (/private/var/folders/66/6_p15z9j15b6gmgyc19g42cw0000gn/T/sbt_5b6a73b1/simple/build.sbt:8)
[info] [error]      Did you mean compile:scalaSource ?
```

using `val base = (scalaSource in Compile).value` gives me a weird `ClassNotFoundException`.

Also what do you mean by outdated operators? I'm not using any deprecated operators as far as I can see.

hseeberger replied:
Strangely our comments are no longer put underneath the respective code, hence it's hard for me to understand your questions.

What I was trying to say is that you should only use the `:=` operator, not `<<=` and friends.

britter replied:
Your comment was about the check task.

I understand that I have to change 

```
TaskKey[Unit]("check") <<= scalaSource in Compile map { (scalaSource) =>
```

to 

```
val checkFileContents = taskKey[Unit]("Verify file contents match the expected contents")
checkFileContents := {
```

But how do I access the value of `scalaSource` from inside the task definitions without usting the `<<=` operator? I tried it but got the errors above.

hseeberger replied:
You can just use any other setting or task key and call `value`: `scalaSource.in(Compile).value`

britter replied:
Ah! I was missing the `in(Compile)`. Thank you. I'll change it.

hseeberger commented on src/sbt-test/sbt-header/simple/project/plugin.sbt:
```diff
@@ -0,0 +1,7 @@
+{
```
Replace with

```
val pluginVersion = sys.props
  .get("plugin.version")
  .getOrElse(sys.error("Sys prop plugin.version must be defined!"))

addSbtPlugin("de.heikoseeberger" % "sbt-fresh" % pluginVersion)
```

britter commented on src/sbt-test/sbt-header/simple/build.sbt:
```diff
@@ -0,0 +1,21 @@
+TaskKey[Unit]("check") <<= scalaSource in Compile map { (scalaSource) =>
+
+  val actualPath = (scalaSource / "Main.scala").toString
+  val expectedPath = (scalaSource / "Main.scala_expected").toString
```
build.sbt only defines a header mapping for scala files. So no, this file won't be affected. Putting the expected files alongside the actual files makes it easier to select them inside the check task.

britter commented:
I think I've addressed all of your comments. Do we want to add some more checks for example for other header formats or for the shebang handling?

I'll definitely need the scripted tests when I start working on #30.

hseeberger commented:
Thanks, mate!
